[Feature] 장바구니 CRUD 구현 및 기타 필드,response 수정

HaenaCho01 opened the pull request:
## <완료>
- 장바구니 메뉴 추가
- 장바구니 조회
- 특정 주문 수량 감소
- 특정 주문 수량 추가
- 특정 주문 취소
- 전체 주문 취소
- 장바구니 주문 완료
- 주문 내역 조회

## <기존 협의한 내용과 바뀐 내용 -> 확인 및 추가 논의 후 수정 필요>
- 장바구니 Response에 가게 이름 추가
- 장바구니 상품 Response에 상품 이름 추가
- 주문 시 PUT 메소드, request List 받지 않고 요구사항 받도록 추가
    - 장바구니 주문 수정을 특정 주문 수량 감소/추가로 나눈 후 api 수정 -> 주문 시 받은 List와 일일이 비교하며 주문으로 넘어가면 로직이 복잡해 일단 이렇게 구현했는데 어떤 방식이 더 나은지 의논 필요
- ERD 상 장바구니 필드에 총액이 존재해 해당 금액이 변경될 때마다 수정되도록 메소드 수정
    - 필드에 넣지 않고 조회 시 response에 반환될 때 계산되어 표시되도록 하는 건 어떤 지 논의 필요

## <와이어프레임 및 프론트 분들 요청사항에 맞게 추가 수정>
- 장바구니, 주문 조회 시 response 'createdDate', 'modifiedDate' 추가
- 상품 필드 'description' 추가
- 상품 response 'storeName' 추가
- 가게 상세 조회 response 'products' 추가

HaenaCho01 committed:
[Init] 가게 API 관련 패키지 및 파일 생성

HaenaCho01 committed:
[Update] common 패키지 예외처리 패키지 추가 및 Timestamped 추가

HaenaCho01 committed:
[Update] 가게 기본 CRUD 구현

- s3, status에 관한 내용 의논 및 수정 필요

HaenaCho01 committed:
[Update] 가게 기본 CRUD 수정

<완료>
- 가게-사용자 관계 추가
- 상태 메소드 entity로 이동 및 public 사용
- 가게 전체조회 메소드명 수정(storeList->stores)
- 응답에 id 추가
- repository에서 가게 찾기 public 수정(상품 서비스에서 찾기 위함)

<할일>
- s3, status에 관한 내용 의논 및 수정 필요

HaenaCho01 committed:
[Init] 상품 기본 CRUD 기초 구현

<할일>
- s3, 인기도에 관한 내용 의논 및 수정 필요

HaenaCho01 committed:
[Update] 가게, 상품 기본 CRUD 개선

<완료>
- 삭제 메소드 컨트롤러 오타 수정 (put->delete) 완료
- 가게 상태 필드 삭제 완료
- 내 가게 조회 메소드 추가 완료

<할일>
- 가게 및 상품 S3 연결
- 상품 인기도 알고리즘 -> 주문 CRUD 완료 후 마무리 예정 (인기 상품 조회 메소드 추가 필요)

HaenaCho01 committed:
[Update] pull

HaenaCho01 committed:
Merge branch 'proLmpa:main' into main

HaenaCho01 committed:
Merge branch 'proLmpa:main' into main

HaenaCho01 committed:
[Init & Update] 장바구니 CRUD 구현

<완료>
- 장바구니 메뉴 추가
- 장바구니 조회
- 특정 주문 수량 감소
- 특정 주문 수량 추가
- 특정 주문 취소
- 전체 주문 취소
- 장바구니 주문 완료
- 주문 내역 조회

HaenaCho01 committed:
[Update] 장바구니 총액 관련 내용 수정

- ERD에 장바구니 총액이 존재해 해당 금액이 변경될 때마다 수정되도록 메소드 수정
- ERD에 넣지 않고 조회 시 response에 계산되서 표시되도록 하는건 어떤지 논의 필요

HaenaCho01 committed:
[Update] 필드 및 response 수정

- 장바구니, 주문 조회 시 response 'createdDate', 'modifiedDate' 추가
- 상품 필드 'description' 추가
- 상품 response 'storeName' 추가
- 가게 상세 조회 response 'products' 추가

proLmpa commented on src/main/java/com/example/baglemonster/store/dto/StoreDetailResponseDto.java:
```diff
@@ -0,0 +1,48 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class StoreDetailResponseDto {
```
StoreResponseDto와 StoreDetailResponseDto의 구조가 동일한데, 하나로 줄일 수 있을 것 같습니다.
 -- 혹시 구별된 용도가 있나요?

HaenaCho01 replied:
StoreResponseDto는 products를 불러오지 않아서 가게 목록에 반환될 때 표시되는 내용이고, StoreDetailResponseDto는 products를 함께 반환하여 가게 상세 조회시 표시되는 내용입니다!

고민해봤는데 가게 목록 조회시 상품 목록이 필요하지 않은데 데이터를 함께 담는 것은 불필요한 것 같아 해당 내용으로 구분했는데

혹시 중복 제거하면서 해당 구분을 할 수 있는 다른 방법 있을까요?

proLmpa commented on src/main/java/com/example/baglemonster/cart/service/CartService.java:
```diff
@@ -0,0 +1,186 @@
+    @Transactional
+    public void createCart(CartRequestDto cartRequestDto, User user) {
+        // 장바구니 확인 후 가져오기
+        User consumer = userService.findUser(user.getId());
```
findUser() 메서드 하나를 쓰기 위해서 UserService 클래스 전체를 가져오는 게 낭비라고 생각되서 UserService 대신 UserRepository를 생성자 주입으로 가져와서 CartService에도 구현해놓는 건 어떨까 싶습니다.

HaenaCho01 replied:
앗 그렇게 구현했다가 userService로 바꿨었는데 넵 수정해놓겠습니다!

HaenaCho01 replied:
수정 완료했습니다!

proLmpa commented on src/main/java/com/example/baglemonster/cart/service/CartService.java:
```diff
@@ -0,0 +1,186 @@
+    private Cart getCart(User user, CartRequestDto cartRequestDto) {
+        // 장바구니가 이미 존재할 경우
+        if (findCartByUserAndStatus(user) != null) {
+            Cart cart = findCartByUserAndStatus(user);
```
같은 메서드가 두 번 사용되는 건 낭비 같아서

Cart cart = findCartByUserAndStatus(user);
if (cart != null) {
  ...
} else {
 ...
}

로 바꿔도 괜찮을 거 같습니다.

HaenaCho01 replied:
오 네넵 급하게 작성하느라 그런 부분을 놓쳤네요. 감사합니다!

HaenaCho01 replied:
수정 완료했습니다!

proLmpa reviewed (approved):
제가 생각했던 부분과 생각지 못했던 부분까지 전부 구현해주셔서 좋았습니다!

약간의 메서드 중복 호출을 제외하면 완벽했습니다 :D:D

HaenaCho01 committed:
[Update] Store 및 Cart 일부 코드 수정

- 가게 및 장바구니 서비스 클래스에서 findUser 메소드 사용 -> userService가 아닌 userRepository를 통해 사용되도록 수정
- 장바구니 서비스 중복 코드 수정

proLmpa commented:
* 주문 시 PUT 메소드, request List 받지 않고 요구사항 받도록 추가
  수량 추가 / 감소 시마다 프론트에서 백으로 일일이 요청을 다 보내는 것이 비효율적이라고 생각하여 조금 번거롭더라도 상품 리스트를 보내는 것이 낫다고 판단하여 ERD를 그렇게 구성했었습니다.
